net/mwan3/files/usr/sbin/mwan3track

```diff
@@ -136,14 +143,23 @@ main() {
 			if [ $host_up_count -lt $reliability ]; then
 				case "$track_method" in
 					ping)
+						# Ensure that IPv6 address is reset
+						unset ADDR
 						if echo $track_ip | grep -q ':'; then
+							if "$KERNEL_HAS_SK_BUG"; then
+								# Pinging IPv6 hosts with an interface is troublesome on
+								# older kernel version. So get the IP address of the
+								# interface and use that instead.
+								ADDR=$(ip -6 addr ls dev "$DEVICE" | sed -ne '/\/128/d' -e '1,/ *inet6 /s/ *inet6 \([^ \/]*\).* scope global.*/\1/p')
+								[ -z "$ADDR" ] && ADDR=$(ip -6 addr ls dev "$DEVICE" | sed -ne 's/ *inet6 \([^ \/]*\).* scope global.*/\1/p')
+							fi
 							ping_protocol=6
 						fi
 						if [ $check_quality -eq 0 ]; then
-							$PING -$ping_protocol -I $DEVICE -c $count -W $timeout -s $size -t $max_ttl -q $track_ip &> /dev/null
+							$PING -$ping_protocol -I ${ADDR:-$DEVICE} -c $count -W $timeout -s $size -t $max_ttl -q $track_ip &> /dev/null
 							result=$?
 						else
-							ping_result_raw="$($PING -$ping_protocol -I $DEVICE -c $count -W $timeout -s $size -t $max_ttl -q $track_ip 2>/dev/null)"
+							ping_result_raw="$($PING -$ping_protocol -I ${ADDR:-$DEVICE} -c $count -W $timeout -s $size -t $max_ttl -q $track_ip 2>/dev/null)"
```

Comment:
It is not related to the IPv6 tracking issue but you could use a single $PING call both when you need or not the output.
Just ignore ping_result_raw when you don't need it. It is just 4 lines.

Just take a look at '-W' that you use. It is only a timeout before the first answer. After that, it does nothing and ping will wait two RTT (I guess the last received reply RTT). It might not make sense here.

I really like ping '-w' option, specially together with '-c'. It allows me to ask "ping for s seconds and expect n replies" or fail ($?==1). Together with -A, you could even do a crude latency test already inside ping. Something like: "Fail if avg latency is greater than 300ms (3s/10 pings) or a packet is lost":

```
ping -A -w 3 -c 10 www.google.com; echo $
ping -A -w 3 -c 10 www.donwdetector.com.ru; echo $?
```

However, the "lost packet" is not perfect here as a very fast link together with a too tolerant latency threshold will allow ping, if a packet is lost, to wait for two RTT and still receive all required replies before the '-w' timeout. If '-W' worked for each echo, you could use it to wait for more than '-w' timeout and make ping fail when a packet was lost.

Reply:
I don't want to add this change to the openwrt-19.07 branch.
I open for your change in the master branch.
So if you want you can make a P/R so we could discuss and merge it.
I only added the change here because the bug in the kernel has been fixed and the workaround in mwan3 was no longer needed here.